File: src/test/resources/generics/GenericInstanceOf.java
package generics;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.Map;

abstract class StateReplicatorServerMessage {
}

class StateReplicatorFullUpdate<T> extends StateReplicatorServerMessage {
    public int serverSideModCount;
    public T state;

    public StateReplicatorFullUpdate(int serverSideModCount, T state) {
        this.serverSideModCount = serverSideModCount;
        this.state = state;
    }
}

public class GenericInstanceOf {

    public static void test1(Object o) {
        if (o instanceof StateReplicatorFullUpdate) {
            System.out.println("1");
        }
        System.out.println(o instanceof StateReplicatorFullUpdate);
        System.out.println(StateReplicatorFullUpdate.class);
        System.out.println(o instanceof List);
        System.out.println(o instanceof Map);
        System.out.println(o instanceof Set);
        System.out.println(o instanceof ArrayList);
        System.out.println(StateReplicatorFullUpdate.class.isAssignableFrom(o.getClass()));
        System.out.println(List.class.isAssignableFrom(o.getClass()));
        System.out.println(Map.class.isAssignableFrom(o.getClass()));
        System.out.println(Set.class.isAssignableFrom(o.getClass()));
        System.out.println(ArrayList.class.isAssignableFrom(o.getClass()));
    }

    public static void test2(StateReplicatorServerMessage o) {
        if (o instanceof StateReplicatorFullUpdate) {
            System.out.println("1");
        }
        System.out.println(o instanceof StateReplicatorFullUpdate);
        System.out.println(o instanceof StateReplicatorFullUpdate ? "1" : "2");
    }

}

File: src/test/resources/generics/GenericInstanceOf.cs
using System.Collections;
using System.Collections.Generic;

namespace generics {
    internal abstract class StateReplicatorServerMessage {
    }

    internal class StateReplicatorFullUpdate<T> : StateReplicatorServerMessage {
        public int serverSideModCount;

        public T state;

        public StateReplicatorFullUpdate(int serverSideModCount, T state) {
            this.serverSideModCount = serverSideModCount;
            this.state = state;
        }
    }

    public class GenericInstanceOf {
        public static void test1(object o) {
            if (Sharpen.Runtime.InstanceOf(o, typeof(StateReplicatorFullUpdate<>))) {
                System.Console.Out.WriteLine("1");
            }
            System.Console.Out.WriteLine(Sharpen.Runtime.InstanceOf(o, typeof(StateReplicatorFullUpdate<>)));
            System.Console.Out.WriteLine(typeof(StateReplicatorFullUpdate<>));
            System.Console.Out.WriteLine(o is IList);
            System.Console.Out.WriteLine(o is IDictionary);
            System.Console.Out.WriteLine(Sharpen.Runtime.InstanceOf(o, typeof(HashSet<>)));
            System.Console.Out.WriteLine(Sharpen.Runtime.InstanceOf(o, typeof(List<>)));
            System.Console.Out.WriteLine(Sharpen.Runtime.IsAssignableFrom(typeof(StateReplicatorFullUpdate<>), o.GetType()));
            System.Console.Out.WriteLine(Sharpen.Runtime.IsAssignableFrom(typeof(IList), o.GetType()));
            System.Console.Out.WriteLine(Sharpen.Runtime.IsAssignableFrom(typeof(IDictionary), o.GetType()));
            System.Console.Out.WriteLine(Sharpen.Runtime.IsAssignableFrom(typeof(HashSet<>), o.GetType()));
            System.Console.Out.WriteLine(Sharpen.Runtime.IsAssignableFrom(typeof(List<>), o.GetType()));
        }

        public static void test2(StateReplicatorServerMessage o) {
            if (Sharpen.Runtime.InstanceOf(o, typeof(StateReplicatorFullUpdate<>))) {
                System.Console.Out.WriteLine("1");
            }
            System.Console.Out.WriteLine(Sharpen.Runtime.InstanceOf(o, typeof(StateReplicatorFullUpdate<>)));
            System.Console.Out.WriteLine(Sharpen.Runtime.InstanceOf(o, typeof(StateReplicatorFullUpdate<>)) ? "1" : "2");
        }
    }
}